l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol

```diff
@@ -0,0 +1,560 @@
+  uint256 public immutable LIFETIME_IN_ROUNDS;
+  uint256 public immutable EXECUTION_DELAY_IN_ROUNDS;
+
+  event VoteCast(Round indexed round, address indexed proposer);
+  event RoundExecuted(Round indexed round, uint256 slashCount);
+
+  constructor(
+    address _instance,
+    ISlasher _slasher,
+    uint256 _slashingUnit,
+    uint256 _quorum,
+    uint256 _roundSize,
+    uint256 _committeeSize,
+    uint256 _roundSizeInEpochs,
+    uint256 _lifetimeInRounds,
+    uint256 _executionDelayInRounds
+  ) {
+    INSTANCE = _instance;
+    SLASHER = _slasher;
+    SLASHING_UNIT = _slashingUnit;
+    QUORUM = _quorum;
+    ROUND_SIZE = _roundSize;
+    COMMITTEE_SIZE = _committeeSize;
+    ROUND_SIZE_IN_EPOCHS = _roundSizeInEpochs;
+    LIFETIME_IN_ROUNDS = _lifetimeInRounds;
+    EXECUTION_DELAY_IN_ROUNDS = _executionDelayInRounds;
+
+    // TODO: Add missing require checks for input parameters
+
+    require(_quorum > 0, Errors.ConsensusSlashingProposer__QuorumMustBeGreaterThanZero());
+    require(_slashingUnit > 0, Errors.ConsensusSlashingProposer__SlashingUnitMustBeGreaterThanZero());
+    require(
+      _lifetimeInRounds > _executionDelayInRounds,
+      Errors.ConsensusSlashingProposer__LifetimeMustBeGreaterThanExecutionDelay(
+        _lifetimeInRounds, _executionDelayInRounds
+      )
+    );
+  }
+
+  /**
+   * @notice Vote for slashing proposals
+   * @param _votes The encoded votes for slashing
+   * @param _sig Signature from the current proposer
+   */
+  function vote(bytes calldata _votes, Signature calldata _sig) external {
+    Slot slot = getCurrentSlot();
+    Round round = computeRound(slot);
+
+    // Get the current proposer from the rollup - only they can submit votes
+    address proposer = getCurrentProposer();
+
+    // Verify signature includes slot to prevent replay attacks across slots and rounds
+    // The proposer must sign the specific combination of slot + votes
+    bytes32 voteHash = keccak256(abi.encodePacked(Slot.unwrap(slot), _votes));
+    require(_sig.verify(proposer, voteHash), Errors.ConsensusSlashingProposer__InvalidSignature());
+
+    // Each byte encodes 2 validators (4 bits each), so we need half the total validators in bytes
+    uint256 expectedLength = COMMITTEE_SIZE * ROUND_SIZE_IN_EPOCHS / 2;
+    require(
+      _votes.length == expectedLength,
+      Errors.ConsensusSlashingProposer__InvalidVoteLength(expectedLength, _votes.length)
+    );
+
+    // Get the round data for the current round
+    RoundData storage roundData = loadRoundData(round, Round.unwrap(round));
+
+    // Check if a vote has already been cast in the current slot
+    require(
+      Slot.unwrap(roundData.lastVoteSlot) < Slot.unwrap(slot),
+      Errors.ConsensusSlashingProposer__VoteAlreadyCastInCurrentSlot(slot)
+    );
+
+    // Store the vote for this round and update round data
+    // QUESTION: Does this trigger multiple separate storage writes? Can we merge them somehow?
```

Comment:
Right now it would because of the sizes. If you make them samller though, and keep track of the variable size input in a different map you could probably do 
```solidity
roundData = RoundData({});
```
Would help you on the `lastVoteSlot´ and `voteCount`, the `votes` still sepate.